src/nnet3/nnet-cudnn-simple-component.cc

```diff
@@ -0,0 +1,798 @@
+  WriteToken(os, binary, "<FilterZStride>");
+  WriteBasicType(os, binary, strides[2]);
+  WriteToken(os, binary, "<FilterXUpscale>");
+  WriteBasicType(os, binary, upscales[0]);
+  WriteToken(os, binary, "<FilterYUpscale>");
+  WriteBasicType(os, binary, upscales[1]);
+  WriteToken(os, binary, "<FilterZUpscale>");
+  WriteBasicType(os, binary, upscales[2]);
+  WriteToken(os, binary, "<InputVectorization>");
+  WriteBasicType(os, binary, static_cast<int32>(input_vectorization_));
+  WriteToken(os, binary, "<FilterParams>");
+  filter_params_.Write(os, binary);
+  WriteToken(os, binary, "<BiasParams>");
+  bias_params_.Write(os, binary);
+  WriteToken(os, binary, "<IsGradient>");
+  WriteBasicType(os, binary, is_gradient_);
+  WriteToken(os, binary, "</CuDNN3DConvolutionComponent>");
+}
+
+void CuDNN3DConvolutionComponent::SetZero(bool treat_as_gradient) {
+  if (treat_as_gradient) {
+    SetActualLearningRate(1.0);
+    is_gradient_ = true;
+  }
+  filter_params_.SetZero();
+  bias_params_.SetZero();
+}
+
+BaseFloat CuDNN3DConvolutionComponent::DotProduct(const UpdatableComponent &other_in) const {
+  const CuDNN3DConvolutionComponent *other =
+    dynamic_cast<const CuDNN3DConvolutionComponent*>(&other_in);
+  return TraceMatMat(filter_params_, other->filter_params_, kTrans) +
+    VecVec(bias_params_, other->bias_params_);
+}
+
+std::string CuDNN3DConvolutionComponent::Info() const {
+  int32 numDimensions;
+  int32 pad_dims[kConvolutionDimension_];
+  int32 stride_dims[kConvolutionDimension_];
+  int32 upscale_dims[kConvolutionDimension_];
+  cudnnConvolutionMode_t mode;
+  cudnnDataType_t float_type;
+  CUDNN_SAFE_CALL(cudnnGetConvolutionNdDescriptor(conv_desc_,
+                                                  kConvolutionDimension_,
+                                                  &numDimensions,
+                                                  pad_dims,
+                                                  stride_dims,
+                                                  upscale_dims,
+                                                  &mode,
+                                                  &float_type)
+                  );
+  KALDI_ASSERT(float_type == cudnn::GetDataType());
+  int32 filter_dims[CUDNN_DIM_MAX];
+  CUDNN_SAFE_CALL(cudnnGetFilterNdDescriptor(filter_desc_,
+                                             kConvolutionDimension_,
+                                             &float_type,
+                                             &numDimensions,
+                                             filter_dims)
+                  );
+  KALDI_ASSERT(float_type == cudnn::GetDataType());
+
+  std::ostringstream stream;
+  stream << UpdatableComponent::Info()
+         << ", input-x-dim=" << input_x_dim_
+         << ", input-y-dim=" << input_y_dim_
+         << ", input-z-dim=" << input_z_dim_
+         << ", filt-x-dim=" << filter_dims[2]
+         << ", filt-y-dim=" << filter_dims[1]
+         << ", filt-z-dim=" << filter_dims[0]
```

Comment:
I'm not following the details of this pull request; I just want to note
that it's Vijay who will be the one to decide to merge this, not me.


On Tue, May 10, 2016 at 11:43 PM, Daniel Galvez <notifications@github.com>
wrote:

> In src/nnet3/nnet-cudnn-simple-component.cc
> <https://github.com/kaldi-asr/kaldi/pull/770#discussion_r62788283>:
>
> > +  CUDNN_SAFE_CALL(cudnnGetFilterNdDescriptor(filter_desc_,
> > +                                             kConvolutionDimension_,
> > +                                             &float_type,
> > +                                             &numDimensions,
> > +                                             filter_dims)
> > +                  );
> > +  KALDI_ASSERT(float_type == cudnn::GetDataType());
> > +
> > +  std::ostringstream stream;
> > +  stream << UpdatableComponent::Info()
> > +         << ", input-x-dim=" << input_x_dim_
> > +         << ", input-y-dim=" << input_y_dim_
> > +         << ", input-z-dim=" << input_z_dim_
> > +         << ", filt-x-dim=" << filter_dims[2]
> > +         << ", filt-y-dim=" << filter_dims[1]
> > +         << ", filt-z-dim=" << filter_dims[0]
>
> That's not true in this case. CUDNN likes to put the "last" dimension
> (typically, the one with lowest stride, stride = 1) into the 0 index of the
> arrays that it returns from function calls. I'm not entirely sure why you
> are saying this.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/770/files/4001fe5047f981ab44e6f693c0a302fdb4f13e5c#r62788283>
>